### PR TITLE
implement StandardController and expose it to JS as Controller.Standard.*

### PR DESCRIPTION
**Pull request by ZappoMan**

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0986f86c349417fc61a2598043bc9fbaf803e8d7 available at: https://gist.github.com/41ca6a8a2408d8ce4b73<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6055.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6055.dmg

File: libraries/input-plugins/src/input-plugins/StandardController.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    };
+    proxy->resetDeviceBindings = [this, &mapper] () -> bool {
+        mapper.removeAllInputChannelsForDevice(_deviceID);
+        this->assignDefaultInputMapping(mapper);
+        return true;
+    };
+    mapper.registerStandardDevice(proxy);
+}
+
+void StandardController::assignDefaultInputMapping(UserInputMapper& mapper) {
```
Comment (jherico) [inline libraries/input-plugins/src/input-plugins/StandardController.cpp]: I don't think having content in this function makes sense.  All of the `assignDefaultInputMapping` functions are replicating work that will be done by the new mapping system.

Comment (ZappoMan) [inline libraries/input-plugins/src/input-plugins/StandardController.cpp]: agreed - in the long run... we will convert this to JSON.. but I left the old style for now to match the other InputDevices so we can repair them all at once.

File: libraries/input-plugins/src/input-plugins/StandardController.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
+//
+//  StandardController.h
+//  input-plugins/src/input-plugins
+//
+//  Created by Stephen Birarda on 2014-09-23.
```
Comment (jherico) [inline libraries/input-plugins/src/input-plugins/StandardController.h]: fix header

File: libraries/input-plugins/src/input-plugins/StandardController.cpp [hunk 3]
```diff
@@ -0,0 +1,156 @@
+//
+//  StandardController.cpp
+//  input-plugins/src/input-plugins
+//
+//  Created by Stephen Birarda on 2014-09-23.
```
Comment (jherico) [inline libraries/input-plugins/src/input-plugins/StandardController.cpp]: fix head

File: libraries/input-plugins/src/input-plugins/StandardController.h [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+        STANDARD_CONTROLLER_BUTTON_DPAD_UP,
+        STANDARD_CONTROLLER_BUTTON_DPAD_DOWN,
+        STANDARD_CONTROLLER_BUTTON_DPAD_LEFT,
+        STANDARD_CONTROLLER_BUTTON_DPAD_RIGHT,
+
+        STANDARD_CONTROLLER_BUTTON_LEFTSHOULDER,
+        STANDARD_CONTROLLER_BUTTON_RIGHTSHOULDER,
+    };
+
```
Comment (jherico) [inline libraries/input-plugins/src/input-plugins/StandardController.h]: missing pose channels

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9667a103aa6295a36aaafd2bb62889c7732bbc2d available at: https://gist.github.com/41ca6a8a2408d8ce4b73<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6055.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6055.dmg

File: libraries/input-plugins/src/input-plugins/StandardController.h [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+#include <qvector.h>
+
+#include "InputDevice.h"
+
+class StandardController : public QObject, public InputDevice {
+    Q_OBJECT
+    Q_PROPERTY(QString name READ getName)
+
+public:
+    enum StandardControllerAxisChannel {
```
Comment (jherico) [inline libraries/input-plugins/src/input-plugins/StandardController.h]: I suggest using the constants in https://github.com/jherico/hifi/blob/controllers/libraries/input-plugins/src/input-plugins/StandardControls.h

We should also start the work of getting rid of the bisected axes.

File: libraries/input-plugins/src/input-plugins/UserInputMapper.h [hunk 6]
```diff
@@ -238,11 +242,19 @@ class UserInputMapper : public QObject, public Dependency {
...
+
 signals:
     void actionEvent(int action, float state);
 
 
 protected:
+    void registerStandardDevice();
+    uint16 _standardDeviceID = 0;
+    DeviceProxy::Pointer _standardDevice;
+    StandardController* _standardController = nullptr;
```
Comment (jherico) [inline libraries/input-plugins/src/input-plugins/UserInputMapper.h]: recommend StandardController::Pointer (i.e. an alias for std::unique_ptr or std::shared_ptr)

File: libraries/input-plugins/src/input-plugins/StandardController.h (hunk 5 above)
Comment (ZappoMan) [inline libraries/input-plugins/src/input-plugins/StandardController.h]: ok - I'll do that work once we've merged the controllers branch

**Comment by hifi-gustavo** [comment]
Build console output for commit: 312bbf6167871c4020e52e5a27cab630df6fc8f0 available at: https://gist.github.com/41ca6a8a2408d8ce4b73<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6055.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6055.dmg

File: libraries/input-plugins/src/input-plugins/StandardController.h (hunk 4 above)
Comment (ZappoMan) [inline libraries/input-plugins/src/input-plugins/StandardController.h]: fixed

**Comment by hifi-gustavo** [comment]
Build console output for commit: 21a473035e2b60d7f854f1abe432cf6e2cd7c586 available at: https://gist.github.com/41ca6a8a2408d8ce4b73<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6055.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6055.dmg

**Pull request merged**